nexus_motion_planner/src/test_motion_plan_cache.cpp

```diff
@@ -0,0 +1,1165 @@
+  different_plan_req.start_state.joint_state.position.at(2) -= 5.05;
+  different_plan_req.goal_constraints
+  .at(0).joint_constraints.at(0).position -= 6.05;
+  different_plan_req.goal_constraints
+  .at(0).joint_constraints.at(1).position += 7.05;
+  different_plan_req.goal_constraints
+  .at(0).joint_constraints.at(2).position -= 8.05;
+
+  /// RobotTrajectory
+
+  // Plan
+  auto plan = get_dummy_panda_plan();
+
+  // Plan with no frame_id in its trajectory header
+  auto empty_frame_plan = plan;
+  empty_frame_plan.joint_trajectory.header.frame_id = "";
+
+  auto different_plan = plan;
+  different_plan.joint_trajectory.points.at(0).positions.at(0) = 999;
+  different_plan.joint_trajectory.points.at(0).positions.at(1) = 999;
+  different_plan.joint_trajectory.points.at(0).positions.at(2) = 999;
+
+  // Test Utils
+
+  std::string prefix;
+
+  // Checks ====================================================================
+
+  // Initially empty
+  prefix = "test_motion_plans.empty";
+
+  check_and_emit(
+    cache->count_plans(g_robot_name) == 0,
+    prefix, "Plan cache initially empty");
+
+  check_and_emit(
+    cache->fetch_all_matching_plans(
+      *move_group, g_robot_name, plan_req, 999, 999).empty(),
+    prefix, "Fetch all plans on empty cache returns empty");
+
+  check_and_emit(
+    cache->fetch_best_matching_plan(
+      *move_group, g_robot_name, plan_req, 999, 999) == nullptr,
+    prefix, "Fetch best plan on empty cache returns nullptr");
+
+  // Put plan empty frame
+  //
+  // Plan must have frame in joint trajectory, expect put fail
+  prefix = "test_motion_plans.put_plan_empty_frame";
+  double execution_time = 999;
+  double planning_time = 999;
+
+  check_and_emit(
+    !cache->put_plan(
+      *move_group, g_robot_name, plan_req, empty_frame_plan,
+      execution_time, planning_time, false),
+    prefix, "Put empty frame plan, no delete_worse_plans, not ok");
+
+  check_and_emit(
+    cache->count_plans(g_robot_name) == 0, prefix, "No plans in cache");
+
+  // Put plan req empty frame
+  //
+  // Plan request must have frame in workspace, expect put fail
+  prefix = "test_motion_plans.put_plan_req_empty_frame";
+  execution_time = 999;
+  planning_time = 999;
+
+  check_and_emit(
+    !cache->put_plan(
+      *move_group, g_robot_name, empty_frame_plan_req, plan,
+      execution_time, planning_time, false),
+    prefix, "Put empty frame req plan, no delete_worse_plans, not ok");
+
+  check_and_emit(
+    cache->count_plans(g_robot_name) == 0, prefix, "No plans in cache");
+
+  // Put first, no delete_worse_plans
+  prefix = "test_motion_plans.put_first";
+  execution_time = 999;
+  planning_time = 999;
+
+  check_and_emit(
+    cache->put_plan(
+      *move_group, g_robot_name, plan_req, plan,
+      execution_time, planning_time, false),
+    prefix, "Put first valid plan, no delete_worse_plans, ok");
+
+  check_and_emit(
+    cache->count_plans(g_robot_name) == 1, prefix, "One plan in cache");
```

Comment:
nit: Add a check for different robot name, ensuring that it returns no result. Might be overkill to do this check everytime we check the cache, so I think checking when the cache is "most populated" is good enough.

Reply:
776c5fdbefa8443ca9a73b87c17ae22d24d710de